File: bench/headless_chrome/src/bench.rs
use anyhow::Result;
use headless_chrome::protocol::cdp::Page::CaptureScreenshotFormatOption;
use headless_chrome::{Browser, LaunchOptions, Tab};
use serde_json::json;
use std::ffi::OsStr;
use std::path::PathBuf;
use std::thread::sleep;
use std::time::{Duration, Instant};

const ITERS: usize = 20;
const WARMUP_ITERS: usize = 3;
const DELAY_CYCLE_MS: &[u64] = &[200, 210, 220, 230, 240];

#[derive(Clone)]
struct Stats {
    median: f64,
    p10: f64,
    n: usize,
    note: Option<&'static str>,
}

fn chrome_path() -> PathBuf {
    if let Ok(p) = std::env::var("CHROME_PATH") {
        let trimmed = p.trim();
        if !trimmed.is_empty() {
            return PathBuf::from(trimmed);
        }
    }
    let home = std::env::var("HOME").expect("HOME must be set or CHROME_PATH must be provided");
    PathBuf::from(format!(
        "{home}/.cache/puppeteer/chrome/linux-131.0.6778.204/chrome-linux64/chrome"
    ))
}

fn median(mut xs: Vec<f64>) -> f64 {
    xs.sort_by(|a, b| a.partial_cmp(b).unwrap());
    xs[xs.len() / 2]
}

fn p10(mut xs: Vec<f64>) -> f64 {
    xs.sort_by(|a, b| a.partial_cmp(b).unwrap());
    xs[(xs.len() as f64 * 0.1) as usize]
}

fn stats(xs: Vec<f64>) -> Stats {
    Stats {
        median: median(xs.clone()),
        p10: p10(xs),
        n: ITERS,
        note: None,
    }
}

fn stats_with_note(xs: Vec<f64>, note: &'static str) -> Stats {
    Stats {
        median: median(xs.clone()),
        p10: p10(xs),
        n: ITERS,
        note: Some(note),
    }
}

fn print_stats(name: &str, s: &Stats) {
    match s.note {
        Some(note) => println!(
            "{name:24} median={:.2}ms p10={:.2}ms (n={}) [{note}]",
            s.median, s.p10, s.n
        ),
        None => println!(
            "{name:24} median={:.2}ms p10={:.2}ms (n={})",
            s.median, s.p10, s.n
        ),
    }
}

fn urlencode(s: &str) -> String {
    let mut out = String::new();
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(b as char)
            }
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

fn selector_gap_html() -> String {
    "<html><body><script>\
        window.__injectedAt = null;\
        setTimeout(() => {\
          const d = document.createElement('div'); d.id = 'target';\
          document.body.appendChild(d);\
          window.__injectedAt = performance.now();\
        }, 200);\
        </script></body></html>"
        .to_string()
}

fn networkidle_static_html() -> String {
    "<!doctype html><html><body>networkidle-static</body></html>".to_string()
}

fn networkidle_deferred_html() -> String {
    r#"<!doctype html><html><body><script>
setTimeout(() => fetch('data:text/plain,deferred'), 250);
</script></body></html>"#
        .to_string()
}

fn wait_for_function_html(delay_ms: u64) -> String {
    format!(
        r#"<!doctype html><html><body><script>
window.__condValue = false;
window.__condAt = null;
setTimeout(() => {{
    window.__condValue = true;
    window.__condAt = performance.now();
}}, {delay_ms});
</script></body></html>"#
    )
}

fn click_when_enabled_html(delay_ms: u64) -> String {
    format!(
        r#"<!doctype html><html><body>
<button id="btn" disabled>click me</button>
<script>
window.__enabledAt = null;
window.__clickedAt = null;
document.getElementById('btn').addEventListener('click', () => {{
    window.__clickedAt = performance.now();
}});
setTimeout(() => {{
    document.getElementById('btn').disabled = false;
    window.__enabledAt = performance.now();
}}, {delay_ms});
</script></body></html>"#
    )
}

fn stats_to_json(s: &Stats) -> serde_json::Value {
    json!({
        "median": s.median,
        "p10": s.p10,
        "n": s.n,
        "note": s.note,
    })
}

fn launch_once() -> Result<Browser> {
    let extra_args: Vec<&OsStr> = vec![
        OsStr::new("--disable-gpu"),
        OsStr::new("--disable-dev-shm-usage"),
    ];
    let options = LaunchOptions {
        headless: true,
        sandbox: false,
        path: Some(chrome_path()),
        args: extra_args,
        ..Default::default()
    };
    Ok(Browser::new(options)?)
}

fn poll_until_true(tab: &Tab, expr: &str) -> Result<()> {
    let deadline = Instant::now() + Duration::from_secs(10);
    loop {
        let ready = tab
            .evaluate(expr, false)?
            .value
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        if ready {
            return Ok(());
        }
        if Instant::now() >= deadline {
            anyhow::bail!("timed out waiting for expression: {expr}");
        }
        sleep(Duration::from_millis(50));
    }
}

fn bench_networkidle(tab: &Tab, html: &str) -> Result<Stats> {
    let url = format!("data:text/html,{}", urlencode(html));
    for _ in 0..WARMUP_ITERS {
        tab.navigate_to(&url)?.wait_until_navigated()?;
    }
    let mut xs = Vec::with_capacity(ITERS);
    for _ in 0..ITERS {
        let t = Instant::now();
        tab.navigate_to(&url)?.wait_until_navigated()?;
        xs.push(t.elapsed().as_secs_f64() * 1000.0);
    }
    Ok(stats(xs))
}

fn main() -> Result<()> {
    let mut cold = Vec::new();
    for _ in 0..5 {
        let t = Instant::now();
        let browser = launch_once()?;
        cold.push(t.elapsed().as_secs_f64() * 1000.0);
        drop(browser);
        sleep(Duration::from_millis(500));
    }
    let launch = Stats {
        median: median(cold.clone()),
        p10: p10(cold),
        n: 5,
        note: None,
    };
    print_stats("launch_chrome", &launch);

    let browser = launch_once()?;

    let mut np = Vec::with_capacity(ITERS);
    for _ in 0..ITERS {
        let t = Instant::now();
        let tab = browser.new_tab()?;
        np.push(t.elapsed().as_secs_f64() * 1000.0);
        let _ = tab.close(false);
    }
    let new_page = stats(np);
    print_stats("new_page", &new_page);

    let tab = browser.new_tab()?;
    let mut gt = Vec::with_capacity(ITERS);
    for _ in 0..ITERS {
        let t = Instant::now();
        tab.navigate_to("about:blank")?.wait_until_navigated()?;
        gt.push(t.elapsed().as_secs_f64() * 1000.0);
    }
    let goto_about_blank = stats_with_note(gt, "wait_until_navigated uses networkAlmostIdle");
    print_stats("goto_about_blank", &goto_about_blank);

    let mut ss = Vec::with_capacity(ITERS);
    for _ in 0..ITERS {
        let t = Instant::now();
        let _ = tab.capture_screenshot(CaptureScreenshotFormatOption::Png, None, None, true)?;
        ss.push(t.elapsed().as_secs_f64() * 1000.0);
    }
    let screenshot = stats(ss);
    print_stats("screenshot", &screenshot);

    let mut ev = Vec::with_capacity(ITERS);
    for _ in 0..ITERS {
        let t = Instant::now();
        let _ = tab.evaluate("document.title", false)?;
        ev.push(t.elapsed().as_secs_f64() * 1000.0);
    }
    let evaluate = stats(ev);
    print_stats("evaluate", &evaluate);

    let selector_url = format!("data:text/html,{}", urlencode(&selector_gap_html()));
    let mut selector_gaps = Vec::with_capacity(ITERS);
    for _ in 0..ITERS {
        tab.navigate_to(&selector_url)?.wait_until_navigated()?;
        let _ = tab.wait_for_element("#target")?;
        let returned_at: f64 = tab
            .evaluate("performance.now()", false)?
            .value
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let injected_at: f64 = tab
            .evaluate("window.__injectedAt", false)?
            .value
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        selector_gaps.push(returned_at - injected_at);
    }
    let wait_for_selector_gap =
        stats_with_note(selector_gaps, "wait_for_element built-in polling");
    print_stats("wait_for_selector_gap", &wait_for_selector_gap);

    let networkidle_static = bench_networkidle(&tab, &networkidle_static_html())?;
    print_stats("networkidle_static", &networkidle_static);

    let networkidle_deferred_250 = bench_networkidle(&tab, &networkidle_deferred_html())?;
    print_stats("networkidle_deferred_250", &networkidle_deferred_250);

    let mut wait_for_function_gaps = Vec::with_capacity(ITERS);
    for i in 0..ITERS {
        let delay = DELAY_CYCLE_MS[i % DELAY_CYCLE_MS.len()];
        let url = format!(
            "data:text/html,{}",
            urlencode(&wait_for_function_html(delay))
        );
        tab.navigate_to(&url)?.wait_until_navigated()?;
        poll_until_true(&tab, "!!window.__condValue")?;
        let returned_at: f64 = tab
            .evaluate("performance.now()", false)?
            .value
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let cond_at: f64 = tab
            .evaluate("window.__condAt", false)?
            .value
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        wait_for_function_gaps.push(returned_at - cond_at);
    }
    let wait_for_function_gap = stats_with_note(wait_for_function_gaps, "manual poll @50ms");
    print_stats("wait_for_function_gap", &wait_for_function_gap);

    let mut click_when_enabled_gaps = Vec::with_capacity(ITERS);
    for i in 0..ITERS {
        let delay = DELAY_CYCLE_MS[i % DELAY_CYCLE_MS.len()];
        let url = format!(
            "data:text/html,{}",
            urlencode(&click_when_enabled_html(delay))
        );
        tab.navigate_to(&url)?.wait_until_navigated()?;
        poll_until_true(&tab, "!document.getElementById('btn').disabled")?;
        tab.find_element("#btn")?.click()?;
        let clicked_at: f64 = tab
            .evaluate("window.__clickedAt", false)?
            .value
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let enabled_at: f64 = tab
            .evaluate("window.__enabledAt", false)?
            .value
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        click_when_enabled_gaps.push(clicked_at - enabled_at);
    }
    let click_when_enabled_gap = stats_with_note(click_when_enabled_gaps, "manual poll @50ms");
    print_stats("click_when_enabled_gap", &click_when_enabled_gap);

    println!(
        "RESULTS_JSON {}",
        json!({
            "library": "headless_chrome",
            "metrics": {
                "launch_chrome": stats_to_json(&launch),
                "new_page": stats_to_json(&new_page),
                "goto_about_blank": stats_to_json(&goto_about_blank),
                "screenshot": stats_to_json(&screenshot),
                "evaluate": stats_to_json(&evaluate),
                "wait_for_selector_gap": stats_to_json(&wait_for_selector_gap),
                "networkidle_static": stats_to_json(&networkidle_static),
                "networkidle_deferred_250": stats_to_json(&networkidle_deferred_250),
                "wait_for_function_gap": stats_to_json(&wait_for_function_gap),
                "click_when_enabled_gap": stats_to_json(&click_when_enabled_gap),
            }
        })
    );

    Ok(())
}

File: bench/puppeteer/bench.ts
// Puppeteer side of the parity bench.
//
// Run:
//   node bench.ts
//   bun bench.ts
const puppeteer = require('puppeteer');
const { performance } = require('perf_hooks');

const CHROME_PATH = process.env.CHROME_PATH
  || `${process.env.HOME}/.cache/puppeteer/chrome/linux-131.0.6778.204/chrome-linux64/chrome`;
const ITERS = 20;
const WARMUP_ITERS = 3;
const DELAY_CYCLE_MS = [200, 210, 220, 230, 240];

function median(xs) {
  const sorted = [...xs].sort((a, b) => a - b);
  return sorted[Math.floor(sorted.length / 2)];
}

function p10(xs) {
  const sorted = [...xs].sort((a, b) => a - b);
  return sorted[Math.floor(sorted.length * 0.1)];
}

function stats(xs, note = null) {
  return { median: median(xs), p10: p10(xs), n: xs.length, note };
}

function printStats(name, s) {
  const base = `${name.padEnd(24)} median=${s.median.toFixed(2)}ms p10=${s.p10.toFixed(2)}ms (n=${s.n})`;
  console.log(s.note ? `${base} [${s.note}]` : base);
}

async function timed(fn) {
  const t0 = performance.now();
  await fn();
  return performance.now() - t0;
}

function selectorGapHtml() {
  return `<html><body><script>
    window.__injectedAt = null;
    setTimeout(() => {
      const d = document.createElement('div'); d.id = 'target';
      document.body.appendChild(d);
      window.__injectedAt = performance.now();
    }, 200);
  </script></body></html>`;
}

function networkidleStaticHtml() {
  return '<!doctype html><html><body>networkidle-static</body></html>';
}

function networkidleDeferredHtml() {
  return `<!doctype html><html><body><script>
    setTimeout(() => fetch('data:text/plain,deferred'), 250);
  </script></body></html>`;
}

function waitForFunctionHtml(delayMs) {
  return `<!doctype html><html><body><script>
    window.__condValue = false;
    window.__condAt = null;
    setTimeout(() => {
      window.__condValue = true;
      window.__condAt = performance.now();
    }, ${delayMs});
  </script></body></html>`;
}

function clickWhenEnabledHtml(delayMs) {
  return `<!doctype html><html><body>
    <button id="btn" disabled>click me</button>
    <script>
      window.__enabledAt = null;
      window.__clickedAt = null;
      document.getElementById('btn').addEventListener('click', () => {
        window.__clickedAt = performance.now();
      });
      setTimeout(() => {
        document.getElementById('btn').disabled = false;
        window.__enabledAt = performance.now();
      }, ${delayMs});
    </script>
  </body></html>`;
}

async function benchNetworkIdle(page, html) {
  const dataUrl = 'data:text/html,' + encodeURIComponent(html);
  for (let i = 0; i < WARMUP_ITERS; i++) {
    await page.goto(dataUrl, { waitUntil: 'networkidle0' });
  }
  const xs = [];
  for (let i = 0; i < ITERS; i++) {
    xs.push(await timed(() => page.goto(dataUrl, { waitUntil: 'networkidle0' })));
  }
  return stats(xs);
}

(async () => {
  const launchArgs = ['--no-sandbox', '--disable-gpu', '--disable-dev-shm-usage'];

  const cold = [];
  for (let i = 0; i < 5; i++) {
    const t = performance.now();
    const browser = await puppeteer.launch({
      headless: 'new',
      executablePath: CHROME_PATH,
      args: launchArgs,
    });
    cold.push(performance.now() - t);
    await browser.close();
    await new Promise((resolve) => setTimeout(resolve, 500));
  }
  const launch = stats(cold);
  printStats('launch_chrome', launch);

  const browser = await puppeteer.launch({
    headless: 'new',
    executablePath: CHROME_PATH,
    args: launchArgs,
  });

  const newPageSamples = [];
  for (let i = 0; i < ITERS; i++) {
    newPageSamples.push(await timed(async () => {
      const page = await browser.newPage();
      await page.close();
    }));
  }
  const newPage = stats(newPageSamples);
  printStats('new_page', newPage);

  const page = await browser.newPage();
  await page.goto('about:blank', { waitUntil: 'load' });

  const gotoSamples = [];
  for (let i = 0; i < ITERS; i++) {
    gotoSamples.push(await timed(() => page.goto('about:blank', { waitUntil: 'load' })));
  }
  const gotoAboutBlank = stats(gotoSamples);
  printStats('goto_about_blank', gotoAboutBlank);

  const screenshotSamples = [];
  for (let i = 0; i < ITERS; i++) {
    screenshotSamples.push(await timed(() => page.screenshot()));
  }
  const screenshot = stats(screenshotSamples);
  printStats('screenshot', screenshot);

  const evaluateSamples = [];
  for (let i = 0; i < ITERS; i++) {
    evaluateSamples.push(await timed(() => page.evaluate(() => document.title)));
  }
  const evaluate = stats(evaluateSamples);
  printStats('evaluate', evaluate);

  const selectorUrl = 'data:text/html,' + encodeURIComponent(selectorGapHtml());
  const selectorGaps = [];
  for (let i = 0; i < ITERS; i++) {
    await page.goto(selectorUrl, { waitUntil: 'load' });
    await page.waitForSelector('#target');
    const returnedAt = await page.evaluate(() => performance.now());
    const injectedAt = await page.evaluate(() => window.__injectedAt);
    selectorGaps.push(returnedAt - injectedAt);
  }
  const waitForSelectorGap = stats(selectorGaps);
  printStats('wait_for_selector_gap', waitForSelectorGap);

  const networkidleStatic = await benchNetworkIdle(page, networkidleStaticHtml());
  printStats('networkidle_static', networkidleStatic);

  const networkidleDeferred250 = await benchNetworkIdle(page, networkidleDeferredHtml());
  printStats('networkidle_deferred_250', networkidleDeferred250);

  const waitForFunctionGaps = [];
  for (let i = 0; i < ITERS; i++) {
    const delay = DELAY_CYCLE_MS[i % DELAY_CYCLE_MS.length];
    const dataUrl = 'data:text/html,' + encodeURIComponent(waitForFunctionHtml(delay));
    await page.goto(dataUrl, { waitUntil: 'load' });
    await page.waitForFunction('window.__condValue === true', { polling: 'raf', timeout: 10000 });
    const returnedAt = await page.evaluate(() => performance.now());
    const condAt = await page.evaluate(() => window.__condAt);
    waitForFunctionGaps.push(returnedAt - condAt);
  }
  const waitForFunctionGap = stats(waitForFunctionGaps);
  printStats('wait_for_function_gap', waitForFunctionGap);

  const clickWhenEnabledGaps = [];
  for (let i = 0; i < ITERS; i++) {
    const delay = DELAY_CYCLE_MS[i % DELAY_CYCLE_MS.length];
    const dataUrl = 'data:text/html,' + encodeURIComponent(clickWhenEnabledHtml(delay));
    await page.goto(dataUrl, { waitUntil: 'load' });
    await page.locator('#btn').click();
    const clickedAt = await page.evaluate(() => window.__clickedAt);
    const enabledAt = await page.evaluate(() => window.__enabledAt);
    clickWhenEnabledGaps.push(clickedAt - enabledAt);
  }
  const clickWhenEnabledGap = stats(clickWhenEnabledGaps);
  printStats('click_when_enabled_gap', clickWhenEnabledGap);

  console.log(`RESULTS_JSON ${JSON.stringify({
    library: 'puppeteer',
    metrics: {
      launch_chrome: launch,
      new_page: newPage,
      goto_about_blank: gotoAboutBlank,
      screenshot,
      evaluate,
      wait_for_selector_gap: waitForSelectorGap,
      networkidle_static: networkidleStatic,
      networkidle_deferred_250: networkidleDeferred250,
      wait_for_function_gap: waitForFunctionGap,
      click_when_enabled_gap: clickWhenEnabledGap,
    },
  })}`);

  await browser.close();
})().catch((err) => {
  console.error(err);
  process.exit(1);
});
